plugins/scaffolder-backend-module-gitea/src/actions/gitea.ts

```diff
@@ -0,0 +1,366 @@
+      getOptions,
+    );
+  } catch (e) {
+    throw new Error(
+      `Unable to get the repository: ${owner}/${repo} metadata , ${e}`,
+    );
+  }
+  return response;
+};
+
+const checkGiteaOrg = async (
+  config: GiteaIntegrationConfig,
+  options: {
+    owner: string;
+  },
+): Promise<void> => {
+  const { owner } = options;
+  let response: Response;
+  // check first if the org = owner exists
+  const getOptions: RequestInit = {
+    method: 'GET',
+    headers: {
+      ...getGiteaRequestOptions(config).headers,
+      'Content-Type': 'application/json',
+    },
+  };
+  try {
+    response = await fetch(
+      `${config.baseUrl}/api/v1/orgs/${owner}`,
+      getOptions,
+    );
+  } catch (e) {
+    throw new Error(`Unable to get the Organization: ${owner}, ${e}`);
+  }
+  if (response.status !== 200) {
+    throw new Error(
+      `Organization ${owner} do not exist. Please create it first !`,
+    );
+  }
+};
+
+const createGiteaProject = async (
+  config: GiteaIntegrationConfig,
+  options: {
+    projectName: string;
+    owner?: string;
+    description: string;
+  },
+): Promise<void> => {
+  const { projectName, description, owner } = options;
+
+  /*
+    Several options exist to create a repository using either the user or organisation
+    User: https://gitea.com/api/swagger#/user/createCurrentUserRepo
+    Api: URL/api/v1/user/repos
+    Remark: The user is the username defined part of the backstage integration config for the gitea URL !
+
+    Org: https://gitea.com/api/swagger#/organization/createOrgRepo
+    Api: URL/api/v1/orgs/${org_owner}/repos
+    This is the default scenario that we support currently
+  */
+  let response: Response;
+
+  const postOptions: RequestInit = {
+    method: 'POST',
+    body: JSON.stringify({
+      name: projectName,
+      description,
+    }),
+    headers: {
+      ...getGiteaRequestOptions(config).headers,
+      'Content-Type': 'application/json',
+    },
+  };
+  try {
+    response = await fetch(
+      `${config.baseUrl}/api/v1/orgs/${owner}/repos`,
+      postOptions,
+    );
+  } catch (e) {
+    throw new Error(`Unable to create repository, ${e}`);
+  }
+  if (response.status !== 201) {
+    throw new Error(
+      `Unable to create repository, ${response.status} ${
+        response.statusText
+      }, ${await response.text()}`,
+    );
+  }
+};
+
+const generateCommitMessage = (
+  config: Config,
+  commitSubject?: string,
+): string => {
+  const changeId = crypto.randomBytes(20).toString('hex');
+  const msg = `${
+    config.getOptionalString('scaffolder.defaultCommitMessage') || commitSubject
+  }\n\nChange-Id: I${changeId}`;
+  return msg;
+};
+
+/**
+ * Checks if the provided function can be executed within a specific period of time limit.
+ * @param fn
+ * @param timeLimit
+ */
+function checkDurationLimit(fn: () => void, timeLimit: number): boolean {
+  const startTime = process.hrtime();
+
+  // Call the function
+  fn();
```

Comment:
This construct is odd to me and probably won't work quite as you intended. `checkAvailabilityGiteaRepository` is async, so it will return immediately, not when the actual operation completes. The duration here will probably always be sub-millisecond.

And besides, you aren't really interested in measuring the duration of an infinite loop. You are interested in entering a single loop, that aborts EITHER if you get the abort signal or if it's been retrying for too long. I think I'd suggest deleting `checkDurationLimit` entirely, and instead introducing a time check inside the loop of `checkAvailabilityGiteaRepository` itself.

Reply:
I reviewed the code: 9507bff4f074